File: kbcuekt/kbcuekt/Managers/http/HttpUrlFileUploadManager.swift
//
//  HttpUrlFileTaskManager.swift
//  kbucket
//
//  Created by 김미혜 on 2017. 12. 16..
//  Copyright © 2017년 kikiplus. All rights reserved.
//

import Foundation

class HttpUrlFileUploadManager {
    
    private let TAG : String = "HttpUrlFileUploadManager"
    
    /**
     * 접속할 URL
     */
    private var mURl : String
    
    /**
     * post방식 true, get방식-false
     */
    private var isPost  : Bool = false
    
    /**
     * HTTP 리시브 콜백 메소드 객체
     */
    private var mIHttpReceive : IHttpReceive
    /**
     * action Id
     */
    private var mId : Int
    
    /**
     * 데이터 경계선 문자열
     */
    private var BOUNDARY_STRING : String = "== DATA END ==="
    
    
    /**
     * 생성자
     */
    public init(url : String, post : Bool,  receive : IHttpReceive, id : Int) {
        mURl = url;
        isPost = post;
        mIHttpReceive = receive;
        mId = id;
    }
    
    
    func actionTask(filePath : String, setValue : String, reqValue : String, fileName : String , image : Data){
        let url = URL(string: mURl)
        KLog.d(tag: TAG, msg: "@@ url : " + mURl )
        let lineEnd = "\r\n"
        let twoHyphens = "--"
        let boundary = "*****"
        let mimetype = "image/jpg"
        
        var request = URLRequest(url: url!)
        request.httpMethod = "POST"
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        request.setValue(setValue, forHTTPHeaderField: reqValue)
        request.setValue(fileName, forHTTPHeaderField: fileName)
        var postString = twoHyphens + boundary + lineEnd
        postString += "Content-Disposition: form-data; name=\"uploadFile\";filename=\"" + fileName + "\"" + lineEnd
        postString += "Content-Type: \(mimetype)\r\n\r\n" + lineEnd
        KLog.d(tag: TAG, msg: "@@ url1 postString : " + postString )
        
        let encodedImageData = image.base64EncodedString(options: NSData.Base64EncodingOptions(rawValue: 0))
        postString += encodedImageData
        postString += lineEnd
        KLog.d(tag: TAG, msg: "@@ url2 postString : " + postString )
        request.httpBody = postString.data(using: .utf8)
        
        KLog.d(tag: TAG, msg: "@@ url3 : " + postString )
        
        let task = URLSession.shared.dataTask(with: request, completionHandler : {
            (data : Data?, response : URLResponse?, error : Error?) -> Void in
            
             KLog.d(tag: "ttt", msg: "@@ url4 : "  )
            guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else{
                 self.mIHttpReceive.onHttpReceive(type: ConstHTTP.HTTP_FAIL, actionId: self.mId, data: data! )
                return
            }
           self.mIHttpReceive.onHttpReceive(type: ConstHTTP.HTTP_OK, actionId: self.mId, data: data! )
        })
        task.resume()
    }
    
    /**
     * Map 형식으로 Key와 Value를 셋팅한다.
     *
     * @param key   : 서버에서 사용할 변수명
     * @param value : 변수명에 해당하는 실제 값
     * @return
     */
    func setParams(key : String , value : String) -> String {
        return "Content-Disposition: form-data; name=\"" + key + "\"\r\n\r\n" + value
    }
    
    /**
     * 업로드할 파일에 대한 메타 데이터를 설정한다.
     *
     * @param key      : 서버에서 사용할 파일 변수명
     * @param fileName : 서버에서 저장될 파일명
     * @return
     */
    func setFile(key : String, fileName : String) -> String {
        return "Content-Disposition: form-data; name=\"" + key + "\";filename=\"" + fileName + "\"\r\n"
    }
    
}

File: kbcuekt/kbcuekt/Managers/http/HttpUrlFileUploadManager.java
package momo.kikiplus.com.kbucket.Managers.http;

import android.graphics.Bitmap;
import android.os.AsyncTask;

import java.io.BufferedReader;
import java.io.ByteArrayInputStream;
import java.io.DataOutputStream;
import java.io.FileInputStream;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;

import momo.kikiplus.com.kbucket.Utils.ByteUtils;
import momo.kikiplus.com.kbucket.Utils.KLog;


/**
 * @author grapegirl
 * @version 1.0
 * @Class Name : HttpUrlFileUploadManager
 * @Description : 파일 업로드
 * @since 2015-07-01.
 */
public class HttpUrlFileUploadManager extends AsyncTask<Object, Void, Void> {

    /**
     * 접속할 URL
     */
    private String mURl = null;

    /**
     * HTTP 리시브 콜백 메소드 객체
     */
    private IHttpReceive mIHttpReceive = null;

    /**
     * 데이터 경계선 문자열
     */
    private String BOUNDARY_STRING = "== DATA END ===";
    /**
     * action Id
     */
    private int mId;

    private byte[] mByteArray;

    /**
     * 생성자
     */
    public HttpUrlFileUploadManager(String url, IHttpReceive receive, int id, byte[] bytes) {
        mURl = url;
        mIHttpReceive = receive;
        mId = id;
        mByteArray = bytes;
    }

    @Override
    protected void onPreExecute() {
        super.onPreExecute();
    }

    @Override
    protected Void doInBackground(Object... params) {
        String filePath = (String) params[0];
        String setValue = (String) params[1];
        String reqValue = (String) params[2];
        String fileName = (String) params[3];

        String lineEnd = "\r\n";
        String twoHyphens = "--";
        String boundary = "*****";

        try {

            ByteArrayInputStream byteArrayInputStream = new ByteArrayInputStream(mByteArray);
            URL connectUrl = new URL(mURl);

            // open connection
            HttpURLConnection conn = (HttpURLConnection) connectUrl.openConnection();
            conn.setDoInput(true);
            conn.setDoOutput(true);
            conn.setUseCaches(false);
            conn.setRequestMethod("POST");
            conn.setRequestProperty("Connection", "Keep-Alive");
            conn.setRequestProperty("Content-Type", "multipart/form-data;boundary=" + boundary);
            conn.setRequestProperty(setValue, reqValue);
            conn.setRequestProperty("filename", fileName);

            // write data
            DataOutputStream dos = new DataOutputStream(conn.getOutputStream());
            dos.writeBytes(twoHyphens + boundary + lineEnd);
            dos.writeBytes("Content-Disposition: form-data; name=\"uploadFile\";filename=\"" + fileName + "\"" + lineEnd);
            dos.writeBytes(lineEnd);

            int bytesAvailable = byteArrayInputStream.available();
            int maxBufferSize = 1024;
            int bufferSize = Math.min(bytesAvailable, maxBufferSize);

            byte[] buffer = new byte[bufferSize];
            int bytesRead = byteArrayInputStream.read(buffer, 0, bufferSize);

            // read image
            while (bytesRead > 0) {
                dos.write(buffer, 0, bufferSize);
                bytesAvailable = byteArrayInputStream.available();
                bufferSize = Math.min(bytesAvailable, maxBufferSize);
                bytesRead = byteArrayInputStream.read(buffer, 0, bufferSize);
            }

            dos.writeBytes(lineEnd);
            dos.writeBytes(twoHyphens + boundary + twoHyphens + lineEnd);

            // close streams
            byteArrayInputStream.close();
            dos.flush(); // finish upload...

            // get response
            int ch;
            InputStream is = conn.getInputStream();
            StringBuffer b = new StringBuffer();
            while ((ch = is.read()) != -1) {
                b.append((char) ch);
            }
            String s = b.toString();
            KLog.e("Test", "result = " + s);
            dos.close();
            mIHttpReceive.onHttpReceive(mIHttpReceive.HTTP_OK, mId, s);

        } catch (Exception e) {
            KLog.d("Test", "exception " + e.getMessage());
            mIHttpReceive.onHttpReceive(mIHttpReceive.HTTP_FAIL, mId, this.getClass() + " @@ Exception ");
        }

        return null;
    }


    @Override
    protected void onPostExecute(Void aVoid) {
        super.onPostExecute(aVoid);
    }


    /**
     * Map 형식으로 Key와 Value를 셋팅한다.
     *
     * @param key   : 서버에서 사용할 변수명
     * @param value : 변수명에 해당하는 실제 값
     * @return
     */
    public static String setParams(String key, String value) {
        return "Content-Disposition: form-data; name=\"" + key + "\"\r\n\r\n" + value;
    }

    /**
     * 업로드할 파일에 대한 메타 데이터를 설정한다.
     *
     * @param key      : 서버에서 사용할 파일 변수명
     * @param fileName : 서버에서 저장될 파일명
     * @return
     */
    public static String setFile(String key, String fileName) {
        return "Content-Disposition: form-data; name=\"" + key
                + "\";filename=\"" + fileName + "\"\r\n";
    }
}

